src/renderer_webgpu.cpp

```diff
@@ -0,0 +1,4808 @@
+				m_samplers[numSamplers + ii] = _fsh->m_samplers[ii];
+				bindings[numBindings++] = _fsh->m_textures[ii];
+				bindings[numBindings++] = _fsh->m_samplers[ii];
+			}
+
+			numSamplers += _fsh->m_numSamplers;
+		}
+
+		for (uint8_t stage = 0; stage < BGFX_CONFIG_MAX_TEXTURE_SAMPLERS; ++stage)
+		{
+			if (isValid(m_vsh->m_bindInfo[stage].m_uniform))
+			{
+				m_bindInfo[stage] = m_vsh->m_bindInfo[stage];
+			}
+			else if (NULL != m_fsh && isValid(m_fsh->m_bindInfo[stage].m_uniform))
+			{
+				m_bindInfo[stage] = m_fsh->m_bindInfo[stage];
+				m_bindInfo[stage].m_index += _vsh->m_numSamplers;
+			}
+		}
+
+		m_numSamplers = numSamplers;
+
+		for (uint32_t ii = 0; ii < _vsh->m_numBuffers; ++ii)
+		{
+			m_buffers[ii] = _vsh->m_buffers[ii];
+			bindings[numBindings++] = _vsh->m_buffers[ii];
+		}
+
+		m_numBuffers = _vsh->m_numBuffers;
+
+		BX_CHECK(m_numUniforms + m_numSamplers * 2 + m_numBuffers == numBindings, "");
+
+		wgpu::BindGroupLayoutDescriptor bindGroupDesc;
+		bindGroupDesc.entryCount = numBindings;
+		bindGroupDesc.entries = bindings;
+		m_bindGroupLayout = s_renderWgpu->m_device.CreateBindGroupLayout(&bindGroupDesc);
+
+		bx::HashMurmur2A murmur;
+		murmur.begin();
+		murmur.add(m_numUniforms);
+		murmur.add(m_textures, sizeof(wgpu::BindGroupLayoutEntry) * numSamplers);
+		murmur.add(m_samplers, sizeof(wgpu::BindGroupLayoutEntry) * numSamplers);
+		murmur.add(m_buffers,  sizeof(wgpu::BindGroupLayoutEntry) * m_numBuffers);
+		m_bindGroupLayoutHash = murmur.end();
+	}
+
+	void ProgramWgpu::destroy()
+	{
+		m_vsh = NULL;
+		m_fsh = NULL;
+		if ( NULL != m_computePS )
+		{
+			BX_DELETE(g_allocator, m_computePS);
+			m_computePS = NULL;
+		}
+	}
+
+	void BufferWgpu::create(uint32_t _size, void* _data, uint16_t _flags, uint16_t _stride, bool _vertex)
+	{
+		BX_UNUSED(_stride);
+
+		m_size = _size;
+		m_flags = _flags;
+		m_vertex = _vertex;
+
+		const uint32_t paddedSize = bx::strideAlign(_size, 4);
+
+		bool storage = m_flags & BGFX_BUFFER_COMPUTE_READ_WRITE;
+		bool indirect = m_flags & BGFX_BUFFER_DRAW_INDIRECT;
+
+		wgpu::BufferDescriptor desc;
+		desc.size = paddedSize;
+		desc.usage = _vertex ? wgpu::BufferUsage::Vertex : wgpu::BufferUsage::Index;
+		desc.usage |= (storage || indirect) ? wgpu::BufferUsage::Storage : wgpu::BufferUsage(0);
+		desc.usage |= indirect ? wgpu::BufferUsage::Indirect : wgpu::BufferUsage(0);
+		desc.usage |= wgpu::BufferUsage::CopyDst;
+
+		m_ptr = s_renderWgpu->m_device.CreateBuffer(&desc);
+
+		if(NULL != _data)
+		{
+			if(_size % 4 != 0)
+			{
+				uint8_t* temp = (uint8_t*)BX_ALLOC(g_allocator, paddedSize);
+				bx::memCopy(temp, _data, _size);
+				m_ptr.SetSubData(0, paddedSize, temp);
+				BX_FREE(g_allocator, temp);
+			}
+			else
+			{
+				m_ptr.SetSubData(0, _size, (const uint8_t*)_data);
+			}
+		}
+	}
+
+	void BufferWgpu::update(uint32_t _offset, uint32_t _size, void* _data, bool _discard)
+	{
+		wgpu::CommandEncoder& bce = s_renderWgpu->getBlitCommandEncoder();
```

Comment:
I think because we don't actually want to create them with `CopyDst` nor use `setSubData` when first creating them ? It's already deprecated, in fact ? I'm getting warnings that `setSubData` is deprecated and must fix that already, and that I should use `createBufferMapped`. Should I ?

Reply:
Sorry, I think I see your point now. So for non dynamic buffers, we want to do CreateBufferMapped and fill them right away, and for dynamic buffers we'lls do setSubData for now. Not sure how a staging allocator would work, here the issue is that we don't want latency in the update of what bgfx calls transient and dynamic buffers ?